dds/DCPS/DispatchService.cpp

```diff
@@ -0,0 +1,207 @@
+/*
+ *
+ *
+ * Distributed under the OpenDDS License.
+ * See: http://www.opendds.org/license.html
+ */
+
+#include "DCPS/DdsDcps_pch.h" //Only the _pch include should start with DCPS/
+
+#include "DispatchService.h"
+#include "TimeDuration.h"
+
+#include <ace/Reverse_Lock_T.h>
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+
+namespace OpenDDS {
+namespace DCPS {
+
+DispatchService::DispatchService(size_t count)
+ : cv_(mutex_)
+ , allow_dispatch_(true)
+ , stop_when_empty_(false)
+ , running_(true)
+ , running_threads_(0)
+ , max_timer_id_(LONG_MAX)
+ , pool_(count, run, this)
+{
+}
+
+DispatchService::~DispatchService()
+{
+  shutdown();
+}
+
+void DispatchService::shutdown(bool immediate, EventQueue* const pending)
+{
+  ACE_Guard<ACE_Thread_Mutex> guard(mutex_);
+  allow_dispatch_ = false;
+  stop_when_empty_ = true;
+  running_ = running_ && !immediate; // && with existing state in case shutdown has already been called
+  cv_.notify_all();
+
+  if (pool_.contains(ACE_Thread::self())) {
+    ACE_DEBUG((LM_DEBUG, "(%P|%t) Warning :: DispatchService::shutdown: Contained Thread Attempting To Call Shutdown."));
```

Comment:
DispatchService is not aware of system debug levels, but this should probably follow the dev guidelines and be a full error message, since it's unrecoverable / unreportable.

Reply:
According to the guidlines, this should be conditioned on `log_level >= Error`.